kotlin/authenticate-with-oauth/src/main/java/com/esri/arcgisruntime/sample/authenticatewithoauth/MainActivity.kt

```diff
@@ -0,0 +1,281 @@
+ * (an ArcGIS Online account, for example). This sample utilizes Android WebView to show theOAuth sign-in page in a dialog.
+ */
+class MainActivity : AppCompatActivity(), AuthenticationChallengeHandler {
+
+  companion object {
+    private val TAG = MainActivity::class.java.simpleName
+  }
+
+  // define configuration for OAuth Portal using custom redirect URL to receive code after auth has been granted
+  private val oAuthConfig: OAuthConfiguration by lazy {
+    OAuthConfiguration(
+      getString(R.string.portal_url),
+      getString(R.string.oauth_client_id),
+      "${BuildConfig.APPLICATION_ID}://${getString(R.string.oauth_redirect_host)}"
+    )
+  }
+
+  // instances of Portal and PortalItem to define what is displayed in the map
+  private val portal: Portal by lazy { Portal(getString(R.string.portal_url)) }
+
+  private val portalItem: PortalItem by lazy { PortalItem(portal, getString(R.string.webmap_world_traffic_id)) }
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    setContentView(R.layout.activity_main)
+
+    // setup AuthenticationManager to handle auth challenges
+    AuthenticationManager.setAuthenticationChallengeHandler(this)
+    AuthenticationManager.addOAuthConfiguration(oAuthConfig)
+  }
+
+  override fun onResume() {
+    super.onResume()
+    mapView.resume()
+
+    handleIntent(intent)
+
+    mapView.map = ArcGISMap(portalItem)
+  }
+
+  override fun onPause() {
+    mapView.pause()
+    super.onPause()
+  }
+
+  override fun onDestroy() {
+    mapView.dispose()
+    super.onDestroy()
+  }
+
+  /**
+   * Attempt to handle the Intent received by the Activity
+   * If the Intent contains an authorization code, store this in the SharedPreferences
+   *
+   * @param intent the Intent to handle
+   */
+  private fun handleIntent(intent: Intent?) {
+    intent?.authCode?.let { code ->
+      sharedPreferences.putAuthCode(code)
+    }
+  }
+
+  /**
+   * Function for handling authentication challenges.
+   *
+   * If an authorization code exists in the SharedPreferences, it's likely that the user is partially through the OAuth flow
+   * and we need to try to obtain the token by performing a OAuthTokenCredentialRequest.
+   *
+   * If there is not an authorization code in the SharedPreferences, it's likely that this is the user's first attempt
+   * at OAuth. So begin OAuth flow.
+   *
+   * @param authenticationChallenge the authentication challenge to handle
+   * @return the AuthenticationChallengeResponse indicating which action to take
+   */
+  override fun handleChallenge(authenticationChallenge: AuthenticationChallenge?): AuthenticationChallengeResponse {
+    authenticationChallenge?.let { authChallenge ->
+
+      try {
+        // if SharedPreferences has an auth code, we've likely just been through the OAuth flow and now have an auth code
+        // we can use to request a new access token
+        sharedPreferences.authCode?.let {
```

Comment:
This is not needed. There is major bug in this sample, that reloads the Portal in onResume() after the user has successfully provided the credential, which causes handleChallenge to be called without finishing the previous OAuth challenge. HandleChallenge() should be called only once per AuthenticationChallenge. This Sample here calls it every time a redirect happens back to this activity.

Also saving auth code in sharedPrefrences is a security flaw and is totally unnecessary for a successful OAuth login workflow.

Reply:
I'd wouldn't call it a major bug, moving that line out of `onResume()` and into `onCreate()` may fix it.

In what way is saving the auth code in `sharedPreferences` a security flaw? It has to be persisted somewhere to be used and since it's cleared after it's used I see no issue with it. Given the fact that the auth code is temporary and the auth code is useless without the secret, I don't see an issue with it at all.

Reply:
>I'd wouldn't call it a major bug, moving that line out of onResume() and into onCreate() may fix it.          

From what I see it will not fix it. 

The flow of the Authentication should be, 
- the user tries to load a Portal, gets challenged with an `OAUTH_CREDENTIAL_CHALLENGE`  in handleChallange(), 
- the app redirects user to the OAuth authentication Portal where the user provides credential, 
- the authenticating portal redirects back to the App with auth code,
-  the App requests for OAuthTokenCredential using the authCode. 

What this sample does is,
-  the user tries to load a Portal, gets challenged with an `OAUTH_CREDENTIAL_CHALLENGE`  in handleChallange(), 
- checks if authcode present in sharedpreferences, if not, then the app redirects user to the OAuth authentication Portal where the user provides credential, 
- the authenticating portal redirects back to the App with auth code,
- the app caches the authcode in SharedPreferences, 
- Loads Portal again, 
- gets challenged with an `OAUTH_CREDENTIAL_CHALLENGE`  in handleChallange() checks sharedpreference for authcode, if present the App requests for OAuthTokenCredential using the authCode. 

So the sample relies on a hack to reload the portal so it can handle the OAuth flow in the HandleChallenge(). We cannot promote this pattern to our users, to tell them that they have to call Load() on a portal twice to complete a successful OAuth authentication.

>In what way is saving the auth code in sharedPreferences a security flaw? It has to be persisted somewhere to be used and since it's cleared after it's used I see no issue with it. Given the fact that the auth code is temporary and the auth code is useless without the secret, I don't see an issue with it at all.     

Persisting any code related to authentication on the client side should be avoided as much as possible cause if you do you are leaving room for vulnerability. You as an app developer might not see a problem with persisting the auth code, but as an api dev we have to push for the right practices/patterns that are secure. 

>It has to be persisted somewhere to be used and since it's cleared after it's used I see no issue with it.       

The only reason why you are persisting the authcode is because you are loading the portal twice to handle the logic in HandleChallenge(). Which is not the right pattern to promote to the users anyway. It is not a requirement to persist the authcode to be able to use it.

Reply:
In order to perform the `OAuthTokenCredentialRequest` when coming back from after the user has entered their credentials we'd need to block using the `CountdownLatch` as you have in the sample you'd provided. Without blocking the thread we need to attempt to load the Portal to fire the `handleChallenge` call. As an app developer, we regularly rely upon APIs to make callbacks in an asynchronous manner to not block any of the threads and allow the UI experience to continue as designed. If we're manually blocking a thread we have not created app side, I wouldn't like to promote this behaviour.

I have not called `load()` on the Portal, if the `MapView` does this for me, it's likely in `mapView.resume()`, I'd expect this behaviour as we should be calling this method during `onResume()` of the app.

I can remove the persisting to `SharedPreferences` and rely just on the `Intent` having the auth token, but again, this is persisted in runtime in the `Activity` `Intent`. There's no getting around the fact that the auth token must persist somewhere, either in runtime memory or elsewhere dependant on the workflow. A leaked auth token with the correct security principals applied to it server side will only be valid for a short period of time. Facebook is 10 minutes if I recall correctly.

Reply:
> I have not called load() on the Portal, if the MapView does this for me, it's likely in mapView.resume(), I'd expect this behaviour as we should be calling this method during onResume() of the app.   

This is the call that loads the map 
`mapView.map = ArcGISMap(portalItem)` 

```Sets the ArcGISMap to display in this MapView. By default, the MapView does not contain an ArcGISMap, and one must be set by calling setMap, which then asynchronously loads the map by calling Loadable.loadAsync on the ArcGISMap.```
http://android.esri.com:8080/docs/arcgis-android/publicAPIRelease/WebJavadoc/com/esri/arcgisruntime/mapping/view/MapView.html#setMap(com.esri.arcgisruntime.mapping.ArcGISMap)

Moving this from onResume() to onCreate() will not make a difference. You are reloading the `Map` which loads the `Portal` which issues a second challenge in `handleResponse()`.

>As an app developer, we regularly rely upon APIs to make callbacks in an asynchronous manner to not block any of the threads and allow the UI experience to continue as designed. If we're manually blocking a thread we have not created app side, I wouldn't like to promote this behaviour.     

Well In order to not do it you have put in place a hack that `Reloads` the Map and `Portal` leading to additional requests. I don't think it is acceptable to reload the `Portal` when the previous `Challenge` is not completed just to get the control back in `handleChallenge()`. 

>I can remove the persisting to SharedPreferences and rely just on the Intent having the auth token, but again, this is persisted in runtime in the Activity Intent. There's no getting around the fact that the auth token must persist somewhere, either in runtime memory or elsewhere dependant on the workflow. A leaked auth token with the correct security principals applied to it server side will only be valid for a short period of time. Facebook is 10 minutes if I recall correctly.      

Having the authenticationCode in Activity Intent is OK as that is how we get it back from the Browser.

Reply:
@PuneetEsri 
> Well In order to not do it you have put in place a hack that Reloads the Map and Portal leading to additional requests. I don't think it is acceptable to reload the Portal when the previous Challenge is not completed just to get the control back in handleChallenge().

How is it a hack if we've returned the `AuthenticationChallengeResponse` in `handleChallenge()`? The previous challenge is completed and ended with `AuthenticationChallengeResponse.Action.CANCEL`.  I believe it's a hack to block the thread. In all my years as an Android App developer I have yet to come across an API that requires a user to block a thread in a callback.

Reply:
After having a discussion with @gunt0001 , it's been decided to re-implement the lock and block the thread. It was noted that that the thread can remain blocked if the user fails to return to the app.